File: problem_067.rs
// -*- compile-command: "rustc -o problem_067_rs problem_067.rs" -*-
// Copyright (c) 2016 Michael Caldwell
use std::fs::File;
use std::io::Read;

fn main() {
    println!("Project Euler - Problem 67:");
    println!("Find the maximum total from top to bottom of the triangle contained in the file, problem_067.txt.\n");

    let mut triangle: [[u32; 100]; 100] = [[0; 100]; 100];

    // Import data from file
    let mut data = String::new();
    let mut f = File::open("problem_067.txt").unwrap();
    f.read_to_string(&mut data).unwrap();

    // Split data into vector
    let data_input: Vec<&str> = data.split_whitespace().collect();

    // Put vector into triangle array
    let mut row = 0;
    let mut col = 0;
    for n in 0..data_input.len() {
        triangle[row][col] = data_input[n].parse::<u32>().unwrap();
        if row == col {
            col = 0;
            row += 1;
        } else {
            col += 1;
        }
    }

    // Solve!
    for y in (0..triangle.len()-1).rev() {
        for x in 0..y+1 {
            if triangle[y+1][x] > triangle[y+1][x+1] {
                triangle[y][x] +=  triangle[y+1][x];
            } else {
                triangle[y][x] +=  triangle[y+1][x+1];
            }
        }
    }

    println!("{}", triangle[0][0]);
}

File: problem_067.swift
#!/usr/bin/xcrun swift
// -*- compile-command: "xcrun -sdk macosx swiftc problem_067.swift -o problem_067_swift" -*-
// Copyright (c) 2016 Michael Caldwell
import Foundation

func debugPrint(a: [[Int]]) -> Void {
    for row in 0 ..< a.count {
        for col in 0 ..< a[row].count {
            if a[row][col] == 0 {
                break
            }
            print(a[row][col], terminator: " ")
        }
        print("")
    }
}

print("Project Euler - Problem 67:")
print("Find the maximum total from top to bottom in a triangle with one-hundred rows.\n")

let debug = false

// Import data from file
let file: FileHandle? = FileHandle(forReadingAtPath: "problem_067.txt")
if file == nil {
    print("Unable to open file.")
    exit(1)
}

// Read all the data
let data = file?.readDataToEndOfFile()

// Close the file
file?.closeFile()

// Convert our data to string
let dataAsString = String(data: data!, encoding: String.Encoding.utf8)

// Separate numbers into array
var array = [[Int]]()
var inputArray = dataAsString!.components(separatedBy:"\n")
for i in 0 ..< inputArray.count - 1{
    var rowArray = [Int]()

    // Split row into array, convert to int and put into rowArray
    let tmp = inputArray[i].components(separatedBy:" ")
    for e in 0 ..< tmp.count {
        rowArray.append(Int(tmp[e])!)
    }

    // Pad rowArray out with 0's so final array is square
    for e in tmp.count...inputArray.count{
        rowArray.append(0)
    }

    // Append new rowArray to final array
    array.append(rowArray)
}

if debug == true {
    debugPrint(a: array)
}

// Solve - Start from bottom
for row in (0 ..< array.count-1).reversed() {
    for col in 0 ..< array.count {
        if array[row+1][col] > array[row+1][col+1] {
            array[row][col] += array[row+1][col]
        } else {
            array[row][col] += array[row+1][col+1]
        }
    }
}

// Print cell 0, 0!
print(array[0][0])

if debug == true {
    debugPrint(a: array)
}
